### DISCUSSION TITLE
Building with `build.rs`

### DISCUSSION THREAD

**Discussion by max-sixty** [wasm-bindgen/wasm-bindgen#3494] [Q&A] [answered]
Is there any facility for building wasm binaries within `build.rs`, rather than calling into `wasm-bindgen`'s CLI?

Over at https://github.com/PRQL/prql/pull/2881 we're hoping to use a `build.rs` file to build our wasm binaries (edit: and JS shims), mostly to take advantage of cargo's caching — `wasm-pack` seems to require a full rebuild even if no files have changed.

We were planning to use https://crates.io/crates/substrate-wasm-builder. That seems to allow building from `build.rs`. It's less focused on JS projects though, so probably need additional CLI calls into `wasm-bindgen`.

I also found https://github.com/rustminded/xtask-wasm/blob/main/src/dist.rs which also looks interesting, but doesn't seem to support building from `build.rs`.

(I also posted this at https://github.com/rustwasm/wasm-pack/issues/916, so feel free to close if that's a better place for it)

Thank you!

**Comment by daxpedda** [comment]
You probably mean how to run `wasm-bindgen` from `build.rs`. Because `wasm-bindgen` itself doesn't build Wasm binaries, it only modifies them and produces a JS shim.

This is indeed possible with [`wasm-bindgen-cli-support`](https://crates.io/crates/wasm-bindgen-cli-support). Which is whats used by `wasm-bindgen-cli` internally and includes all that's necessary to do the same.

**Reply by max-sixty** [reply]
Thanks @daxpedda .

Do you have any thoughts on whether the concept of building from `build.rs` is reasonable? I had a look at the [dependents of wasm-bindgen-cli-support](https://crates.io/crates/wasm-bindgen-cli-support/reverse_dependencies), and they all look to be build tools.

I'm surprised I'm the first person to think that `cargo build` should build the full project, including the JS shims — am I missing something?

**Reply by daxpedda** [reply, marked as answer]
> I'm surprised I'm the first person to think that `cargo build` should build the full project, including the JS shims — am I missing something?

That is the desired goal ofc, unfortunately this requires a dedicated target in Rust, which doesn't currently exist. Something like `wasm32-web`. I'm not entirely sure what you intend to do btw, but you can't build the JS shim from `build.rs` for that same project.

The current flow looks like this:
Rust runs your `build.rs` -> then it compiles your crate -> produces a `.wasm` file -> you run it through `wasm-bindgen-cli`.

You can't let your `build.rs` generate the JS shim and fix you `.wasm` file because it runs before compiling the crate in the first place.
What you can do is do it from another crate.

**Reply by max-sixty** [reply]
>> am I missing something?

> You can't let your build.rs generate the JS shim and fix you .wasm file because it runs before compiling the crate in the first place.

Ah, this is what I was missing! Thank you. So it's more difficult than I thought. 

Thanks!

**Reply by max-sixty** [reply]
FYI — re the "can we do it all from `build.rs`" question — I think that https://crates.io/crates/substrate-wasm-builder is doing this by [creating a new crate](https://github.com/paritytech/substrate/blob/master/utils/wasm-builder/src/wasm_project.rs#L495), compiling it, and then using the artifact as the `.wasm` files. 

So it is indeed possible, but not easy...

**Reply by daxpedda** [reply]
> FYI — re the "can we do it all from `build.rs`" question — I think that https://crates.io/crates/substrate-wasm-builder is doing this by [creating a new crate](https://github.com/paritytech/substrate/blob/master/utils/wasm-builder/src/wasm_project.rs#L495), compiling it, and then using the artifact as the `.wasm` files.

Wow, that's pretty hacky. As far as I can read though they didn't find a nice way to prevent the native build, so it builds twice, once during `build.rs` and then the crate itself.

I think whats missing here is some post-build options for Cargo. Configuring [custom runners](https://doc.rust-lang.org/cargo/reference/config.html#targettriplerunner) works pretty nicely, I've been using [`wasm-server-runner`](https://crates.io/crates/wasm-server-runner) for a while now and it's incredibly convenient.
